Update serving over HTTP docs. (#1813)
graphql/graphql.github.io: src/pages/learn/serving-over-http.mdx

import { Callout } from "nextra/components"

# Serving over HTTP

<p className="learn-subtitle">Respond to GraphQL requests using an HTTP server</p>

The GraphQL specification doesn't require paricular client-server protocols when sending API requests and responses, but HTTP is the most common choice because of its ubiquity. On this page, we'll review some key guidelines to follow when setting up a GraphQL server to operate over HTTP.

Note that the guidelines that follow only apply to stateless query and mutation operations. Visit the [Subscriptions page](/learn/subscriptions) for more information on transport protocols that commonly support these requests.

<Callout type="info">

> The recommendations on this page align with the detailed [HTTP transport specification](https://github.com/graphql/graphql-over-http) currently in development. Though not yet finalized, this draft specification acts as a single source of truth for GraphQL client and library maintainers, detailing how to expose and consume a GraphQL API using an HTTP transport. Unlike the language specification, adherence is not mandatory, but most implementations are moving towards these standards to maximize interoperability.

Reviewer: ```suggestion
The recommendations on this page align with the detailed [GraphQL-over-HTTP specification](https://graphql.github.io/graphql-over-http/draft/) currently in development. Though not yet finalized, this draft specification acts as a single source of truth for GraphQL client and library maintainers, detailing how to expose and consume a GraphQL API using an HTTP transport. Unlike the language specification, adherence is not mandatory, but most implementations are moving towards these standards to maximize interoperability.
```

</Callout>

## API endpoint

> HTTP is commonly associated with REST, which uses "resources" as its core concept. In contrast, GraphQL's conceptual model is an [entity graph](/learn/thinking-in-graphs/). As a result, entities in GraphQL are not identified by URLs. Instead, a GraphQL server operates on a single URL/endpoint, usually `/graphql`, and all GraphQL requests for a given service should be directed at this endpoint.

Reviewer: Love this explanation/wording - very clear and concise! If you want to submit edits to the GraphQL-over-HTTP spec to use nicer wording too, that would be very welcome!

```suggestion
HTTP is commonly associated with REST, which uses "resources" as its core concept. In contrast, GraphQL's conceptual model is an [entity graph](/learn/thinking-in-graphs/). As a result, entities in GraphQL are not identified by URLs. Instead, a GraphQL server operates on a single URL/endpoint, usually `/graphql`, and all GraphQL requests for a given service should be directed to this endpoint.
```

## Where auth happens

> Most modern web frameworks use a pipeline model where requests are passed through a stack of _middleware_, also known as  _filters_ or _plugins_. As the request flows through the pipeline, it can be inspected, transformed, modified, or terminated with a response. GraphQL should be placed after all authentication middleware so that you have access to the same session and user information you would in your HTTP endpoint handlers.

Reviewer: `+other` because if I'm not mistaken, GraphQL is often implemented as an "HTTP endpoint handler"?

```suggestion
Most modern web frameworks use a pipeline model where requests are passed through a stack of _middleware_, also known as _filters_ or _plugins_. As the request flows through the pipeline, it can be inspected, transformed, modified, or terminated with a response. GraphQL should be placed after all authentication middleware so that you have access to the same session and user information you would in your other HTTP endpoint handlers.
```

> After authentication, the server should not make any [authorization](/learn/authorization/) decisions about the request until GraphQL execution begins. Specifically, field-level authorization should be done by the GraphQL schema during [GraphQL's ExecuteRequest()](https://spec.graphql.org/draft/#ExecuteRequest()), allowing for a partial response to be generated if authorization-related errors are raised.

Reviewer: I fear this wording may be slightly misleading - the authorization shouldn't really be done _by_ the GraphQL schema, it should be done by the _business logic_ which the schema calls out to via resolvers. I've attempted to reflect this subtlety with minimal changes:

```suggestion
After authentication, the server should not make any [authorization](/learn/authorization/) decisions about the request until GraphQL execution begins. Specifically, field-level authorization should be enforced by the business logic called from resolvers during [GraphQL's ExecuteRequest()](https://spec.graphql.org/draft/#ExecuteRequest()), allowing for a partial response to be generated if authorization-related errors are raised.
```

## Request format

### Headers

GraphQL clients and servers should support JSON for serialization and may support other formats. Clients should also indicate what media types they support in responses using the `Accept` HTTP header. Specifically, the client should include the `application/graphql-response+json` in the `Accept` header.

If no encoding information is included with this header, then it will be assumed to be `utf-8`. However, the server may respond with an error if a client doesn't provide the `Accept` header with a request.

<Callout type="info">

> The `application/graphql-response+json` is described in the draft GraphQL over HTTP specification. To ensure compatibility, if a client sends a request to a legacy GraphQL server before 1st January 2025, the `Accept` header should also include the `application/json` media type as follows: `application/graphql-response+json; charset=utf-8, application/json; charset=utf-8`

Reviewer: Though spaces after the semicolons are _allowed_, they should be omitted for consistency (and to make reading it as a comma-separated list easier):

```suggestion
The `application/graphql-response+json` is described in the draft GraphQL over HTTP specification. To ensure compatibility, if a client sends a request to a legacy GraphQL server before 1st January 2025, the `Accept` header should also include the `application/json` media type as follows: `application/graphql-response+json;charset=utf-8, application/json;charset=utf-8`
```

</Callout>

### Methods

> Your GraphQL HTTP server must handle the HTTP `POST` method, and may also accept the `GET` method for incoming API requests.

Reviewer: ```suggestion
Your GraphQL HTTP server must handle the HTTP `POST` method for query and mutation operations, and may also accept the `GET` method for query operations.
```

#### `POST` request and body

A standard GraphQL POST request should set `application/json` for its `Content-type` header, and include a JSON-encoded body of the following form:

```json
{
  "query": "...",
  "operationName": "...",
  "variables": { "myVariable": "someValue", ... },
  "extensions": { "myExtension": "someValue", ... }
}
```

> The `query` parameter is required and will contain the source text of the GraphQL document. Note that this parameter may contain documents for query or mutation operations.

Reviewer: Technically the document can contain `subscription` operations too, we just can't execute them. This is useful for when the document contains all of the operations of your application (a mixture of all three operation types), which some people do when using a single trusted document for the entirety of a mobile application.

```suggestion
The `query` parameter is required and will contain the source text of a GraphQL document. Note that the term `query` here is a misnomer: the document may contain any valid GraphQL operations (and associated fragments).
```

> The `operationName`, `variables`, and `extensions` parameters are optional fields. `operationName` is only required if multiple operations are present in the query.

Reviewer: ```suggestion
The `operationName`, `variables`, and `extensions` parameters are optional fields. `operationName` is only required if multiple operations are present in the `query` document.
```

Query variables can be sent as a JSON-encoded string in an additional query parameter called `variables`. If the query contains several named operations, an `operationName` query parameter can be used to control which one should be executed.

#### Choosing an HTTP method

When choosing an HTTP method for a GraphQL request, there are a few points to consider. First, support for HTTP methods other than `POST` will be at the discretion of the GraphQL server, so a client will be limited to the supported verbs.

> Additionally, the `GET` HTTP method may only be used for query operations, so if a client sends a mutation operation in a document then it must use the `POST` method instead.

Reviewer: Subtly incorrect, the document may contain mutations and be sent over GET so long as the chosen operation to execute (via `operationName`) is a query operation.

```suggestion
Additionally, the `GET` HTTP method may only be used for `query` operations, so if a client requests execution of a `mutation` operation then it must use the `POST` method instead.
```

> When servers that do support the `GET` method for query operations, a client may be encouraged to leverage this option to facilitate HTTP caching or edge caching in a content delivery network (CDN). However, because GraphQL documents strings may be quite long for complex operations, the query parameters may exceed the limits that browsers and CDNs impose on URL lengths.

Reviewer: ```suggestion
When servers do support the `GET` method for query operations, a client may be encouraged to leverage this option to facilitate HTTP caching or edge caching in a content delivery network (CDN). However, because GraphQL documents strings may be quite long for complex operations, the query parameters may exceed the limits that browsers and CDNs impose on URL lengths.
```

> In these cases, it would be necessary to use something like _trusted documents_ (sometimes referred to as _persisted queries_) or _automatic persisted queries_, where the server stores hashes of known GraphQL documents and the clients can send a hash as the `query` parameter value in a `GET` request instead of the full source text of the query operation.

Reviewer: This network optimization doesn't require _trusted_ documents: _persisted_ documents is sufficient for this. Different techniques supply the identifier in different ways (the persisted documents RFC is trying to standardize this) so I've avoided being specific here though I have linked to the RFC:

```suggestion
In these cases, a common approach is for the server to store identified GraphQL documents using a technique such as [_persisted documents_](https://github.com/graphql/graphql-over-http/pull/264), _automatic persisted queries_, or _trusted documents_. This allows the client to send a short document identifier instead of the full query text.
```

## Response format

### Headers

The response type should be `application/graphql-response+json` but can be `application/json` to support legacy clients. The server will indicate the media type in the `Content-type` header of the response. It should also indicate the encoding, otherwise `utf-8` will be assumed.

### Body

Regardless of the HTTP method used to send the query and variables, the response should be returned in the body of the request in JSON format. As mentioned in the GraphQL specification, a query might result in some data and some errors, and those should be returned in a JSON object of the form:

```json
{
  "data": { ... },
  "errors": [ ... ],
  "extensions": { ... }
}
```

> If no errors were returned, the `errors` field must not be present in the response. If errors occurred before execution could start, the `errors` field must include these errors and the `data` field must not be present in the response. The `extensions` field is optional and information provided here will be at the discretion of the GraphQL implementation.

Reviewer: @graphql/tsc please confirm I'm not missing something about the second sentence in the original text... I couldn't quite make sense of it?

> If no data is returned, [according to the GraphQL spec](https://spec.graphql.org/October2021/#sec-Data), the `"data"` field should only be included if no errors occurred during execution.

We've instead replaced it with the following, but I want to be sure we've not missed something:

> If errors occurred before execution could start, the `errors` field must include these errors and the `data` field must not be present in the response.

> If no errors were returned, the `errors` field should not be present in the response. If no data is returned, [according to the GraphQL spec](https://spec.graphql.org/October2021/#sec-Data), the `data` field should only be included if no errors occurred before execution. The `extensions` field is optional and information provided here will be at the discretion of the GraphQL implementation.

Reviewer: (I know this is old text.) I do not understand the intent of the second sentence... 

```suggestion
If no errors were returned, the `errors` field must not be present in the response. If errors occurred before execution could start, the `errors` field must include these errors and the `data` field must not be present in the response. The `extensions` field is optional and information provided here will be at the discretion of the GraphQL implementation.
```

You can read more about GraphQL spec-compliant response formats on the [Response page](/learn/response/).

### Status Codes

> If a response contains a `data` key and its value is not `null`, then the server should respond with a `2xx` status code for either the `application/graphql-response+json` or `application/json` media types.

Reviewer: ```suggestion
If a response contains a `data` key and its value is not `null`, then the server should respond with a `2xx` status code for either the `application/graphql-response+json` or `application/json` media types. This is the case even if the response includes errors, since HTTP does not yet have a status code representing "partial success."
```

For validation errors that prevent the execution of a GraphQL operation, the server will typically send a `400` status code, although some legacy servers may return a `2xx` status code when the `application/json` media type is used.

For legacy servers that respond with the `application/json` media type, the use of `4xx` and `5xx` status codes for valid GraphQL requests that fail to execute is discouraged but may be used depending on the implementation. Because of potential ambiguities regarding the source of the server-side error, a `2xx` code should be used with this media type instead.

However, for responses with the `application/graphql-response+json` media type, the server will reply with a `4xx` or `5xx` status code if a valid GraphQL request fails to execute.

## Server implementations

If you use Node.js, we recommend looking at the [list of JS server implementations](/community/tools-and-libraries/?tags=javascript_server).

You can view a [list of servers written in many other languages here](/community/tools-and-libraries/?tags=server).

## Recap

To recap these recommendations for serving GraphQL over HTTP:

> - The server should handle user authentication before a GraphQL request is validated; authorization should happen during request execution against the schema

Reviewer: ```suggestion
- The server should handle user authentication before a GraphQL request is validated; authorization should happen within your business logic during GraphQL request execution
```
